feat : req v2 ui and integrations

kushagrasarathe opened the pull request:
- added req v2 parser
- integrated new api routes for req/charge/payment creation
- status views based on charge and request id's

fixes TASK-8486
fixes TASK-7032
fixes TASK-8032
fixes TASK-8889
fixes TASK-8890

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Enhanced the payment workflow with new components for payment history, payment form, payment confirmation, and payment status views, providing real-time updates and clear transaction details.
	- Introduced a guest login modal, simplifying authentication and improving the overall user experience.
	- Added a user-friendly error handling component for displaying payment-related errors.
	- Implemented a new loading indicator and improved interface elements for better navigation and responsiveness across devices.
	- Added a new `PaymentInfoRow` component for displaying payment details in a structured format.
	- Introduced `ValidationErrorView` to inform users of invalid payment URLs with clear instructions.
	- Added new error classes for enhanced validation error handling.
	- Introduced a `PaymentLayout` component for improved structure in payment-related pages.

- **Bug Fixes**
	- Resolved various validation errors and improved error messaging for better user guidance during payment processes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kushagrasarathe committed:
fix: consider gas calculation wen sending native tokens

kushagrasarathe committed:
feat: payment history ui

kushagrasarathe committed:
feat: basic redux slice setup for payment view

kushagrasarathe committed:
feat: basic dynamic route setup for recipient

kushagrasarathe committed:
feat: chain validation + validation errors

kushagrasarathe committed:
feat: add supported tokens record

kushagrasarathe committed:
feat: url resolver parser

kushagrasarathe committed:
feat: validate token utility

kushagrasarathe committed:
feat: add recipient validation utility

kushagrasarathe committed:
feat: add amount validation

kushagrasarathe committed:
feat: validate payment params utility fn

kushagrasarathe committed:
fix: directory structure

kushagrasarathe committed:
feat: update token addresses

kushagrasarathe committed:
feat: add native token support to chain resolver

kushagrasarathe committed:
feat: add peanut username validation

kushagrasarathe committed:
feat: set chain and token details from req url to token selector

kushagrasarathe committed:
feat: create req and then pay a req

kushagrasarathe committed:
fix: remove fee buffer

kushagrasarathe committed:
fix: revert input view changes

kushagrasarathe committed:
fix: send native tokens fail even wen enough balance

kushagrasarathe committed:
Merge branch 'peanut-wallet-dev' into feat/req-v2-parser

kushagrasarathe committed:
Merge branch 'peanut-wallet-dev' into feat/req-v2-parser

kushagrasarathe committed:
Merge branch 'peanut-wallet-dev' into feat/req-v2-parser

kushagrasarathe committed:
feat: add recipient resolver

kushagrasarathe committed:
feat: requests and chagres types

kushagrasarathe committed:
feat: add requests and charges services

kushagrasarathe committed:
feat: create requests and charges api integration

kushagrasarathe committed:
feat: pay request integration

kushagrasarathe committed:
Merge branch 'peanut-wallet-dev' into feat/req-v2-ui

jjramirezn committed:
fix: toast fail on request link creation

kushagrasarathe committed:
Merge branch 'feat/req-v2-ui' into feat/req-v2-parser

kushagrasarathe committed:
feat: update chain resolver to handle chain specific addreses

kushagrasarathe committed:
feat: update chain validation

kushagrasarathe committed:
feat: add address parser to resolve chain-specific addresses

kushagrasarathe committed:
feat: update url parser to handle chain-specific addresses

kushagrasarathe committed:
fix: charge creation flow

kushagrasarathe committed:
feat: add peanut username validations

kushagrasarathe committed:
fix: add peanut username check before resolving ens

kushagrasarathe committed:
fix: set chain/token values in token selector from url

kushagrasarathe committed:
fix: loading and error states for initial and confirm pay view

kushagrasarathe committed:
feat: render request history

kushagrasarathe committed:
feat: render network/slippage details

kushagrasarathe committed:
feat: request charge status view

kushagrasarathe committed:
feat: legacy request payment

kushagrasarathe committed:
fix: build errors

kushagrasarathe committed:
Merge pull request #677 from peanutprotocol/feat/req-v2-parser

feat: req v2 integrations

jjramirezn commented on src/lib/validation/recipient.ts:
```diff
@@ -0,0 +1,56 @@
+        default:
+            throw new RecipientValidationError('Invalid recipient type')
+    }
+}
```
I agree with this comment but don't think is blocking

jjramirezn commented on src/components/Global/Icon/index.tsx:
```diff
@@ -208,13 +211,11 @@ const icons = {
 type IconProps = {
     className?: string
-    name: IconNameType
+    name: any
```
Agree, any bad

jjramirezn commented on src/app/(mobile-ui)/[...recipient]/error.tsx:
```diff
@@ -0,0 +1,26 @@
+            </Card.Content>
+        </Card>
+    )
+}
```
@kushagrasarathe is this true? blocking if so I agree with changing the name

jjramirezn commented on src/lib/validation/resolvers/recipient-resolver.ts:
```diff
@@ -0,0 +1,36 @@
+    if (isPeanutUsername) {
+        // todo: move to env or constants
+        // append .testvc.eth to resolve Peanut username
+        const peanutEns = `${recipient}.testvc.eth`
```
@kushagrasarathe blocking agree, use env variable

jjramirezn commented on src/app/(mobile-ui)/[...recipient]/page.tsx:
```diff
@@ -0,0 +1,220 @@
+
+                        // if username has no dots, treat as native peanut username
+                        if (!nameToResolve.includes('.')) {
+                            // todo: move to env
```
chore(blocking): do the todo

jjramirezn commented on src/components/Global/Icon/index.tsx:
```diff
@@ -208,13 +211,11 @@ const icons = {
 type IconProps = {
     className?: string
-    name: IconNameType
+    name: any
```
comment: any = bad

kushagrasarathe committed:
fix: pr review nits

jjramirezn commented on src/app/(mobile-ui)/[...recipient]/page.tsx:
```diff
@@ -0,0 +1,220 @@
+                // fetch requests using the resolved address
+                const fetchedRequest = await requestsApi.search({
+                    recipient: recipientAddress,
+                    // todo: add other filters here
```
chore(blocking): do this todo

jjramirezn reviewed (changes requested):
Very big PR, confidence in my review quality: low

Some blocking comments and also guest flow is very important

jjramirezn committed:
feat: multi-device request links history

kushagrasarathe committed:
feat: req pay for guest flow

jjramirezn committed:
fix: dasboad data

jjramirezn committed:
fix: correct user history endpoint

kushagrasarathe committed:
fix: chainId resolve error wen fetching request details
- also updates peanut spin logo for status views
- also fixes incorrect ens resolutions in list item view component
- also fixes incorrect chain id being passed when creating payment
- minor ui changes in charge status view

jjramirezn reviewed (approved):
Approved the changes. Nothing wrong pops out. We will test once merged

jjramirezn committed:
fix: dont trim handle with space

This was causing the validator to hang because the change in value
was not being registered

jjramirezn committed:
refactor: build link in history with new request

kushagrasarathe committed:
feat: parser test cases + error ui component

kushagrasarathe committed:
Merge branch 'peanut-wallet-dev' into feat/req-v2-ui

jjramirezn committed:
fix: signup cta

Hugo0 commented on src/app/(mobile-ui)/layout.tsx:
```diff
@@ -24,11 +18,7 @@ const publicPathRegex = /^\/(request\/pay|claim)/
-    const { user, isFetchingUser } = useAuth()
-    const { handleLogin, isLoggingIn } = useZeroDev()
-    const toast = useToast()
+    const { user } = useAuth()
```
well this is a nice simplifying diff

Hugo0 commented on src/app/(mobile-ui)/layout.tsx:
```diff
@@ -105,55 +95,7 @@ const Layout = ({ children }: { children: React.ReactNode }) => {
-                    </Button>
-                </div>
-            </Modal>
+            <GuestLoginModal />
```
nice!

Hugo0 commented on src/app/[...recipient]/error.tsx:
```diff
@@ -0,0 +1,26 @@
+            <Card.Header>
+                <Card.Title>Something went wrong!</Card.Title>
+                <Card.Description>
+                    {error.message || 'An error occurred while loading the payment page.'}
```
nit: is "the payment page" always accurate? Might be overly specific? or no?

kushagrasarathe replied:
imo it is good, since it's a page where I can pay users or pay a request created by a user, it makes more sense to keep it like this, do you have any other suggestions ?

Hugo0 replied:
oki!

Hugo0 commented on src/app/[...recipient]/layout.tsx:
```diff
@@ -0,0 +1,55 @@
+        <div className="flex h-[100dvh] w-full bg-background">
+            {/* Wrapper div for desktop layout */}
+            <div className="flex w-full">
+                {/* Sidebar - Fixed on desktop */}
```
love the comments

Hugo0 commented on src/app/[...recipient]/loading.tsx:
do we need this component? why not use PeanutLoading directly?

kushagrasarathe replied:
added this to smoothen the ux a bit, it's next.js loading for a particular route and helps show a loading state while all the components in that route are being loaded.

Hugo0 commented on src/app/[...recipient]/page.tsx:
```diff
@@ -0,0 +1,236 @@
+
+export default function PaymentPage({ params }: { params: { recipient: string[] } }) {
+    const dispatch = useAppDispatch()
+    const { currentView, attachmentOptions, resolvedAddress, requestDetails } = usePaymentStore()
```
thoughts on redux vs react context? super out of scope for this PR, but maybe we should revisit this at some point

kushagrasarathe replied:
wdym? revisit as in only keep use one thing out of two, if that's what you mean then yes thats the plan, rn i'm using redux for every new feature we build, ideas is to migrate old ones too to redux gradually, sofar it hasn't worked well due to time constraints tho

Hugo0 commented on src/app/[...recipient]/page.tsx:
```diff
@@ -0,0 +1,236 @@
+            const parsed = parsePaymentURL(params.recipient)
+
+            // if chain is specified but token not, set USDC as default
+            if (parsed.chain && !parsed.token) {
```
again, great comments

Hugo0 commented on src/app/[...recipient]/page.tsx:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    if (isLoadingRequests && !chargeId) {
+        return <PeanutLoading />
```
nit: idk  why we have loading.tsx if we use peanutloading directly?

kushagrasarathe replied:
the transition b/w different views (Initial, confirm) wasn't smooth wen reqs were being fetched, so showing these loading states for those, while loading.tsx is for next route segments

Hugo0 commented on src/app/[...recipient]/page.tsx:
```diff
@@ -0,0 +1,236 @@
+
+    if (isLoadingRequests && !chargeId) {
+        return <PeanutLoading />
+    }
```
1. these 2 statements can be combined
2. might be better to have an explicit isLoadingRequest or isLoadingCharge.  abd not have difficult to parse `|| (chargeId && !currentView)` which is hard to understand without a comment

kushagrasarathe replied:
adding a comment explaing

Hugo0 commented on src/components/Global/FeeDescription/index.tsx:
why is this part of the PR?

kushagrasarathe replied:
included them to use in the confirm view, going forward, we will follow similar ui structure and reuse these components, mainly `InfoRow`

Hugo0 commented on src/components/Global/FlowHeader/index.tsx:
```diff
@@ -12,7 +12,8 @@ interface FlowHeaderProps {
     const { user } = useAuth()
 
-    if (!user) return null
+    // todo: check if this condition removal is affecting other flows
```
?

kushagrasarathe replied:
it's a note, needed to test this change for other flows, forgot to remove the comment, mb :*(

Hugo0 commented on src/components/Global/GuestLoginModal/index.tsx:
```diff
@@ -0,0 +1,68 @@
+            onClose={() => {
+                signInModal.close()
+            }}
+            title={'Sign In with your Peanut Wallet'}
```
nit: lowercase in

Hugo0 commented on src/components/Global/GuestLoginModal/index.tsx:
Do we need this modal? Can't we just redirect to setup? Or is this valuable to have to not have to redirect?

Hugo0 replied:
If we want to keep user in line, wouldnt a redirect to setup with a post-login redirect work better? less code duplication

kushagrasarathe replied:
actually this is good imo, the way i see it is, from ux pov, if i'm on payment page, i want to pay a req, and i'm not logged in, using this modal i've tow options, either i login to my peanut user using my passkey or i connect a external wallet to pay that request

Hugo0 commented on src/components/Payment/History/index.tsx:
```diff
@@ -0,0 +1,55 @@
+        if (type === 'REQUEST' && userRole === 'RECIPIENT') return 'Request paid'
+        if (type === 'CASHOUT') return 'Cash Out'
+        return type
+    }
```
nit: this is never used

Hugo0 commented on src/components/Payment/History/index.tsx:
```diff
@@ -0,0 +1,55 @@
+    return (
+        <div className="space-y-3 border-b border-b-black">
+            <div className="text-base font-semibold">
+                Payment history to <span className="underline">{recipient}</span>
```
Nit: we should have a universal component for identities/accounts that's not just underline: `<span className="underline">{`

Hugo0 commented on src/components/Payment/History/index.tsx:
```diff
@@ -0,0 +1,55 @@
+                                    ? `Paid By: ${entry.senderAccount.username || printableAddress(entry.senderAccount.identifier)} | Status: ${entry.status}`
+                                    : `Status: ${entry.status}`,
+                                subText: new Date(entry.timestamp).toLocaleDateString(),
+                                disableEnsResolution: true,
+                            }}
```
Q: why disable ensResolution? We should resolve the addresses if possible! If we have an ens name for an address, rather show that than the address itself

kushagrasarathe replied:
agree, there's a bug with that, i need to parse the recipient id separately in this case since in this case, otherwise ListItemView would try to parse the weird string like `To: 0cad.... | Status: NEW`, resulting in error, i've this thing tracked on my end, but considering the other issues in hand, had to push this back for a while

Hugo0 commented on src/components/Payment/PaymentForm/index.tsx:
```diff
@@ -0,0 +1,361 @@
+
+    const urlChainName = chain && getReadableChainName(chain?.toString())
+
+    const showPeanutWalletWarning =
```
nit: this is unused

Hugo0 commented on src/components/Payment/PaymentForm/index.tsx:
```diff
@@ -0,0 +1,361 @@
+            if (upperToken === 'ETH') {
+                // Set native token address
+                setSelectedTokenAddress('0x0000000000000000000000000000000000000000')
+            } else {
```
Q: what about native tokens on other chains? e.g. POL on polygon

Hugo0 replied:
Why do we need special behaviour for native eth, cant we get it from indexing supportedtokens w/ upperToken?

Hugo0 commented on src/components/Payment/PaymentForm/index.tsx:
```diff
@@ -0,0 +1,361 @@
+
+        // set token from URL if present
+        if (token) {
+            const upperToken = token.toUpperCase()
```
nit: why upperCase QQ

we only use upperToken for string comparison. just explicitly lowercase (or uppercase) both sides during the string comparison

Hugo0 commented on src/components/Payment/PaymentForm/index.tsx:
```diff
@@ -0,0 +1,361 @@
+                    amount: tokenValue,
+                    currency: 'USD',
+                },
+                baseUrl: window.location.origin,
```
@jjramirezn we also have a NEXT PUBLIC BASE URL variable. Why did we introduce it? can't we use window.location.origin everywhere? and if not, shouldnt this use  the env var?

jjramirezn replied:
This should use the env var for consistency. We added NEXT_PUBLIC_BASE_URL because somethings would break on preview and also to have it available on the next server side

Hugo0 commented on src/services/charges.ts:
```diff
@@ -0,0 +1,65 @@
+import { CreateChargeRequest, PaymentCreationResponse, TCharge, TRequestChargeResponse } from './services.types'
+
+export const chargesApi = {
+    create: async (data: CreateChargeRequest): Promise<TCharge> => {
+        const response = await fetch(`/api/proxy/charges`, {
+            method: 'POST',
+            headers: {
+                'Content-Type': 'application/json',
+            },
+            body: JSON.stringify(data),
+        })
+
+        if (!response.ok) {
+            throw new Error(`Failed to create charge: ${response.statusText}`)
+        }
+
+        return response.json()
+    },
```
Q: Do we need this code for every single call? Can't we DRY a bit here?

Also I'd rather just raise the error message instead of adding a useless "Failed to create charge" infront

kushagrasarathe replied:
wdym, i did not get this

> Do we need this code for every single call?

do you mean, that content type, body, method type specification to be reused somehow ?

Hugo0 replied:
yes, i meant that. it's a very nitpick issue though and maybe bad approach

Hugo0 commented on src/components/Payment/Views/Confirm.payment.view.tsx:
```diff
@@ -0,0 +1,476 @@
+    const [isLoading, setIsLoading] = useState<boolean>(false)
+    const { sendTransactions, checkUserHasEnoughBalance } = useCreateLink()
+    // todo: use redux store for this
+    const [transactionHash, setTransactionHash] = useState<string>('')
```
nit: some unused vars here

Hugo0 commented on src/components/Payment/Views/Error.validation.view.tsx:
```diff
@@ -0,0 +1,38 @@
+                    } else {
+                        router.push('/signin')
+                    }
+                }}
```
/signin leads to a website that says "invalid payment URL" - whats going on here

Hugo0 commented on src/lib/url-parser/constants/tokens.ts:
```diff
@@ -0,0 +1,52 @@
+        },
+        minAmount: '0.01',
+        maxAmount: '1000000',
+    },
```
Issue, Blocking: This is like the 4th place (or more) we now define tokens & chains in (we have in sdk, in bridge supported tokens, in tokenselector, in app, prob in api too, and prob other places too including where we handle squid stuff ). I'm very afraid of so much repetition and how it makes maintainance and debugging much harder. Please give it a think if we can combine this with other places where we maintain tokens and chains, or if we really need a separate registry

Hugo0 commented on src/lib/url-parser/constants/tokens.ts:
```diff
@@ -0,0 +1,52 @@
+            [arbitrum.id]: '0xfd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9',
+            [bsc.id]: '0x55d398326f99059ff775485246999027b3197955',
+        },
+        minAmount: '0.01',
```
Blocking Q: Why the arbitrary minimum amount of 1 cent? And for eth, why the mega arbitrary 20cents minimum amount?

Hugo0 commented on src/lib/url-parser/constants/tokens.ts:
What is the purpose of this file? Do we only support these tokens/chains for request? Is the todo: add more chain addresses actionable? When is that scoped in?

Hugo0 commented on src/lib/url-parser/parsers/address.parser.ts:
If this file references the erc-7828 implementation, I think adding this as a comment would be very helpful

Hugo0 commented on src/lib/url-parser/parsers/address.parser.ts:
```diff
@@ -0,0 +1,55 @@
+    // validate both parts exist
+    if (!user || !chain) {
+        return null
+    }
```
Q, nit: Why not just return user in this case? also is this user or is this address?

e.g. `0x125125512@` or `kushagra.eth@` would qualify here. I think its slightly better to return kushagra.eth as the parsed param, and null for chain, instead of null for everything

Hugo0 commented on src/lib/url-parser/parsers/address.parser.ts:
```diff
@@ -0,0 +1,55 @@
+    // check for ENS names first (including those starting with 0x)
+    if (user.endsWith('.eth')) {
+        return ENS_REGEX.test(user)
+    }
```
Issue, blocking: as previously mentioned, ens names dont have to end with `.eth` - please use justaname resolver for validating ens names

Hugo0 commented on src/lib/url-parser/parsers/address.parser.ts:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    // for ens names, must match regex
+    return ENS_REGEX.test(user)
```
nit: these comments are overly verbose

Hugo0 commented on src/lib/validation/constants/index.ts:
```diff
@@ -0,0 +1,8 @@
+// L1 TLD list based on supported chains
+export const TLDS = ['eth', 'arbitrum', 'optimism', 'base', 'polygon'] as const
+
```
Issue: same comment as with tokens/chains earlier: do we REALLY need yet another place where we define chains? Please please please lets combine these, this is a breeding ground for unintended behaviour and trouble debugging down the line. The simpler, the better, and this PR already adds 3000 lines of code

Hugo0 commented on src/lib/url-parser/parsers/address.parser.ts:
```diff
@@ -0,0 +1,55 @@
+}
+
+// helper function to validate user
+function isValidUser(user: string): boolean {
```
nit: the function name "isValidUser" is potentially confusing imo. We resolve:

- peanut user names (no .TLD needed)
- raw addresses
- arbitrary .eth names

All of these resolve to a _single_ _address_. I think `isValidAddress` or `resolvesToAddress` would be better names. This would also mean that we would have to rename the other existing function that we have, `isAddress`, to a more specific name like `isEVMAddress` like suggested in a previous PR review

Hugo0 commented on src/lib/url-parser/types/token.ts:
```diff
@@ -0,0 +1,9 @@
+    logoURI?: string
+    minAmount?: string
+    maxAmount?: string
+}
```
nit: Im worried about repetition again - we already have types for tokens defined in other places. Why do we need to define again? and what's the reasoning behind having min/max?

Hugo0 commented on src/lib/validation/resolvers/chain-resolver.ts:
```diff
@@ -0,0 +1,178 @@
+}
+
+export function resolveChain(chainIdentifier: string): ResolvedChain {
+    // case 1: chain id (0x[a-fA-F0-9]{1,64})
```
nit: good to specify the particular eip that defines chainid namespace, if any

Hugo0 commented on src/lib/validation/resolvers/chain-resolver.ts:
```diff
@@ -0,0 +1,178 @@
+// normalize chain names using CHAIN_NAME_VARIANTS mapping
+export function normalizeChainName(chainName: string): string {
+    const normalized = chainName.toLowerCase().trim()
+    // check if the normalized string matches any of the variants
```
mini-nit: useless comment imo

Hugo0 commented on src/lib/validation/resolvers/chain-resolver.ts:
```diff
@@ -0,0 +1,178 @@
+
+// normalize chain names using CHAIN_NAME_VARIANTS mapping
+export function normalizeChainName(chainName: string): string {
+    const normalized = chainName.toLowerCase().trim()
```
we trim and lowercase (good) but we don't remove spaces.

Q: What about spaces in the request link? is the expectation someone would write `pay/kushagra@ethereum%20mainnet`?

jjramirezn replied:
I think we should expect or accept spaces in these links. The Idea is they should be human readeable and also human writeable

Hugo0 commented on src/lib/validation/resolvers/chain-resolver.ts:
```diff
@@ -0,0 +1,178 @@
+    const token = SUPPORTED_TOKENS[tokenSymbol.toUpperCase()]
+    if (!token) return false
+
+    return !!token.addresses[Number(chainId)]
```
nit: 
either shorten it like this:
`return token && !!token.addresses[Number(chainId)]`

or, ideally, write it in a way that is easy to parse

```
if (!token) {
return false;
}
if (!token.addresses[Number(chainId)]) {
return false;
}
return true;
```

This last way is more verbose, but easier to understand when reading

Hugo0 commented on src/lib/validation/resolvers/recipient-resolver.ts:
```diff
@@ -0,0 +1,37 @@
+
+    // todo: add support for other ENS TLDs
+    // if its an ENS name (ends with .eth)
+    if (recipient.endsWith('.eth')) {
```
again, .eth not only ens tld

Hugo0 replied:
if it has a dot in the name, try to resolve it with justaname - we discussed thsi

Hugo0 commented on src/lib/url-parser/parser.ts:
```diff
@@ -0,0 +1,184 @@
+import { ParsedURL, RecipientType } from './types/payment'
+
+export function detectRecipientType(recipient: string): RecipientType {
+    if (recipient.endsWith('.eth')) return 'ENS'
```
eth not only ens tld, see other commetns

Hugo0 commented on src/lib/url-parser/parser.ts:
```diff
@@ -0,0 +1,184 @@
+    }
+
+    // check for ENS names
+    if (recipient.endsWith('.eth')) {
```
^

Hugo0 commented on src/lib/validation/resolvers/recipient-resolver.ts:
```diff
@@ -0,0 +1,37 @@
+import { isAddress } from 'viem'
+import { verifyPeanutUsername } from '../recipient'
+
+export async function resolveRecipientToAddress(recipient: string): Promise<string> {
```
nit: here we're resolving recipient to address, but we also have a function called `isValidUser` that is very similar (but with a very different name!). Can we consolidate?

Hugo0 commented on src/lib/validation/resolvers/recipient-resolver.ts:
```diff
@@ -0,0 +1,37 @@
+import { verifyPeanutUsername } from '../recipient'
+
+export async function resolveRecipientToAddress(recipient: string): Promise<string> {
+    // if already an address, return it
```
overly verbose comment, this code is very clear

Hugo0 commented on src/lib/validation/resolvers/recipient-resolver.ts:
```diff
@@ -0,0 +1,37 @@
+        return recipient
+    }
+
+    // todo: add support for other ENS TLDs
```
just use justaname resolver

Hugo0 commented on .env.example:
```diff
@@ -1,3 +1,5 @@
+export NEXT_PUBLIC_JUSTANAME_ENS_DOMAIN=""
```
nit: better to keep vars semantically grouped (e.g. all NEXT_PUBLIC ones together)

Hugo0 commented on src/services/charges.ts:
```diff
@@ -0,0 +1,65 @@
+
+export const chargesApi = {
+    create: async (data: CreateChargeRequest): Promise<TCharge> => {
+        const response = await fetch(`/api/proxy/charges`, {
```
issue, nonblocking: consistency with proxy routes. here you use /api/proxy/charges, but in a different place you use `next_proxy_url`

jjramirezn replied:
agree, also next_proxy_url should be UPPER_CASE

Hugo0 commented on src/lib/validation/amount.ts:
```diff
@@ -0,0 +1,50 @@
+        } catch (error) {
+            throw new AmountValidationError('Invalid amount precision')
+        }
+    }
```
Q: rather throw than round down?

Hugo0 commented on src/lib/validation/chain.ts:
```diff
@@ -0,0 +1,24 @@
+        // convert number to hex string
+        const identifier = typeof chainIdentifier === 'number' ? `0x${chainIdentifier.toString(16)}` : chainIdentifier
+
+        // resolve chain
```
these are very very very very overly verbose comments.

// resolve chain
.... resolvechain()

if you use AI to generate this, please clean up afterwards, don't force a PR review when code is like this

Hugo0 commented on src/lib/validation/resolvers/chain-resolver.ts:
```diff
@@ -0,0 +1,178 @@
+    id: string
+    name: string
+    isTestnet: boolean
+}
```
Issue, blocking: why do we need a `ResolvedChain` type? Why not just use a generic chain type that we use throughout the app? And even if we made a new type, why call it resolvedChain and not type: chain or smth?

please please please lets keep repetition to a minimum. This PR adds 3000 LoC and I worry that a lot of it is unnecessary and surface for bugs

Hugo0 replied:
We only use this type in _one_ place. If that's the case, I'd rather just be explicit about the types in the object, instead of giving it a name like ResolvedChain. Now when i go on chain.ts and hover over the return type of `resolveChain` it says it returns `ResolvedChain` type, which tells me absolutely nothing, and i have to go check out what ResolvedChain type is, which results in bad devx

Hugo0 commented on src/lib/validation/resolvers/chain-resolver.ts:
```diff
@@ -0,0 +1,178 @@
+    },
+    polygon: {
+        variants: ['polygon', 'polygon pos', 'matic', 'polygon mainnet'],
+        supportsNativeEth: false,
```
Q, issue: what is the logic behind supportsNativeEth? Why do we need this? Seems like a potential footgun.

Also as mentioned earlier, we define chains and tokens in sooooooooooo many places. Walletconnect chains, mobula chains, now these chains, its a complete mess. Please make an effort to consolidate if possible

Hugo0 commented on src/lib/validation/recipient.ts:
```diff
@@ -0,0 +1,56 @@
+    switch (type) {
+        case 'ENS':
+            // check if the recipient is a valid ENS name
+            if (!recipient.endsWith('.eth')) {
```
Two issues:
1. as mentioned earlier, more ens TLDs than just .eth
2. why do we need image.png so many places where we check for .eth?
- `src/lib/url-parser` twice
- `address.parser.ts`
- `recipient.ts`
- `chain-resolver.ts` (this one is potentially understandable)
- `recipient-resolver.ts`

This seems like insane repetition to me and i don't think we need it

Hugo0 commented on src/lib/validation/token.ts:
```diff
@@ -0,0 +1,40 @@
+    info: TokenInfo
+    address?: string // optional because ETH is native
+    isNative: boolean
+}
```
issue: same as resolvedChain, this type definition seems nonsensical to me

Hugo0 commented on src/lib/validation/token.ts:
```diff
@@ -0,0 +1,40 @@
+
+export interface ValidatedToken {
+    info: TokenInfo
+    address?: string // optional because ETH is native
```
praise: there's a comment explaining why address is optional, which is good, we should always comment edge cases that are hard to understand. And eth in this case is an edge case. However the more elegant method is to just deal with it so that you catch edge cases - in other cases we've handled the native gas token by giving it the 0x00 address

Hugo0 commented on src/lib/validation/token.ts:
```diff
@@ -0,0 +1,40 @@
+
+    if (chain) {
+        const normalizedChainName = normalizeChainName(chain.name)
+        const isNative = normalizedSymbol === 'ETH' && supportsNativeEth(normalizedChainName)
```
nit: When doing string comparisons, its a good practice to lowercase/uppercase explicitly during the comparison. In this case it says "normalizedSymbol" but if im debugging im not sure, from just one glance, if normalized means its been uppercased or lowercased (alongside trimming etc)

Hugo0 commented on src/redux/types/payment.types.ts:
```diff
@@ -0,0 +1,18 @@
+export type TPaymentView = 'INITIAL' | 'CONFIRM' | 'SUCCESS'
+
+export interface IPaymentState {
+    currentView: 'INITIAL' | 'CONFIRM' | 'SUCCESS'
```
Q: what about payment state where you just reload the page to check the payment status? would that be success?

kushagrasarathe replied:
yes, should have renamed to STATUS, poor code management by me :*(

Hugo0 commented on src/utils/__tests__/url-parser.test.ts:
```diff
@@ -0,0 +1,175 @@
+                chain: '42161',
+            })
+        })
+    })
```
nit: can generate more test cases than just arbitrum

Hugo0 commented on src/utils/__tests__/url-parser.test.ts:
Issue, blocking: no tests for peanut usernames!

Also, test format seems not ideal imo. We should test the full string format as a user would write it - not already with extracted URL params. I would love to test specifically strings like this:

`<baseURL>/konrad@optimism/10usdc`
`/konrad/10`
`konrad/usdc`
`konrad@base/usdc`
`konrad.eth/0.1/eth`

Also, 3000 LoC PR but only 150 LoC for tests seems skewed to me.

Hugo0 reviewed (changes requested):
+3600 Lines of Code is a lot, and I worry we're repeating instead of reusing code that we have for other functions, thus  increasing surface for bugs and making debugging and further maintainance much harder.

kushagrasarathe committed:
fix: use squid-chain-token data for parser + helllaaa lot of refactoring :)

kushagrasarathe committed:
fix: update test cases

kushagrasarathe committed:
Merge branch 'feat/req-v2-ui' of https://github.com/peanutprotocol/peanut-ui into feat/req-v2-ui

kushagrasarathe committed:
fix: use single status component for both charge and request payment status

kushagrasarathe committed:
Merge branch 'peanut-wallet-dev' into feat/req-v2-ui

Hugo0 commented on src/utils/__tests__/url-parser.test.ts:
```diff
@@ -0,0 +1,396 @@
+                    recipientType: 'ENS',
+                    resolvedAddress: '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045',
+                }),
+                chain: expect.objectContaining({ chainId: 1 }),
```
why is chainid number? we represent chainids as strings

kushagrasarathe replied:
actually treating all the params as string for payments, so for consistency kept it like this 👀

Hugo0 commented on src/utils/__tests__/url-parser.test.ts:
```diff
@@ -0,0 +1,396 @@
+            if (recipient.includes('.')) return 'ENS'
+            if (recipient.startsWith('0x')) return 'ADDRESS'
+            return 'USERNAME'
+        },
```
1. this mock function is better than our real function lmao

2. why do we need a mock function? cant we import getrecipientype?

kushagrasarathe replied:
refactored, added it during debugging, forgot to remove 🫠

Hugo0 commented on src/lib/url-parser/parser.consts.ts:
1. why spaces? 
2. this has to be massively expanded no?

Suggestion: we already have different chain names from our diff chain definition/sources (mobula, squid, reown, wagmi, etc). we can just combine them and match on chainid

Hugo0 commented on src/lib/url-parser/parser.ts:
```diff
@@ -0,0 +1,166 @@
+
+    // match number part (including decimals) followed by token part
+    // this regx will match patterns like "0.1usdc", "100eth", "1.5pol"
+    const match = amountString.match(/^(\d*\.?\d*)([a-zA-Z]+)$/)
```
test for this regex?
should ideally have a parset.test.ts file right next to this file, in the same order structure, that tests this regex and parseAmountAndToken

kushagrasarathe replied:
as discussed, will put these in a new separate pr

Hugo0 commented on src/lib/url-parser/parser.ts:
```diff
@@ -0,0 +1,166 @@
+
+    const recipientDetails = await validateAndResolveRecipient(recipient)
+
+    // get all squid chains and tokens
```
1. useless comment
2. two awaits? this sucks

consider if we need two awaits, makes code slow. If we really need, you can parallelize them with promise all

Hugo0 commented on src/lib/url-parser/parser.ts:
```diff
@@ -0,0 +1,166 @@
+    let chain: string | undefined = undefined
+    let chainDetails: (interfaces.ISquidChain & { tokens: interfaces.ISquidToken[] }) | undefined
+
+    // handle multiple @ symbols - take the last one as chain identifier
```
good comment. but what's an example where this could be the case?

kushagrasarathe replied:
i'll add a test case for this 🫡

Hugo0 commented on src/lib/url-parser/parser.ts:
```diff
@@ -0,0 +1,166 @@
+            }
+        }
+
+        // if only amount specified, return with current chain
```
where is current chain coming from?

Hugo0 replied:
should the response here even include a chain in that case? wouldnt it be better to include no chain in the response, and for the caller to decide what to do with the missing chain?

kushagrasarathe replied:
- current chain is the chain details from squid data, based on url param
- ahahha, good catch, mb, should be undefined actually

Hugo0 commented on src/utils/general.utils.ts:
```diff
@@ -347,18 +347,32 @@ export async function resolveFromEnsNameAndProviderUrl(
 
 export async function resolveFromEnsName(ensName: string): Promise<string | undefined> {
-    let resolvedAddress: string | undefined
+    const mainProviderUrl = 'https://mainnet.infura.io/v3/' + infuraApiKey
```
shouldnt infuraapikey be a const and thus uppercase?

kushagrasarathe replied:
it is a const, updating it to be uppercase, this was like this from before ig

Hugo0 reviewed (approved):
APPROVED

kushagrasarathe committed:
fix: nits

kushagrasarathe committed:
fix: fee & slippage calculation

kushagrasarathe committed:
fix: remove wallet header from home page
